eth/gasestimator/gasestimator.go

```diff
@@ -233,3 +239,320 @@ func run(ctx context.Context, call *core.Message, opts *Options) (*core.Executio
+	// other non-fixable conditions
+	var (
+		blockContext = core.NewEVMBlockContext(opts.Header, opts.Chain, nil, opts.Config, opts.State)
+		txContext    = vm.TxContext{
+			Origin:   *tx.Rip7560TransactionData().Sender,
+			GasPrice: tx.GasFeeCap(),
+		}
+
+		dirtyState = opts.State.Copy()
+		evm        = vm.NewEVM(blockContext, txContext, dirtyState, opts.Config, vm.Config{NoBaseFee: true})
+	)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go func() {
+		<-ctx.Done()
+		evm.Cancel()
+	}()
+
+	// Gas Pool is set to half of the maximum possible gas to prevent overflow
+	vpr, err := core.ApplyRip7560ValidationPhases(opts.Config, opts.Chain, &opts.Header.Coinbase, new(core.GasPool).AddGas(math.MaxUint64/2), dirtyState, opts.Header, tx, evm.Config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err != nil {
+		if errors.Is(err, vm.ErrOutOfGas) {
+			return nil, nil, nil // Special case, raise gas limit
+		}
+		return nil, nil, err // Bail out
+	}
+	return vpr, dirtyState, nil
+}
+
+func EstimateRip7560Validation(ctx context.Context, tx *types.Transaction, opts *Options, gasCap uint64) (uint64, error) {
+	// Binary search the gas limit, as it may need to be higher than the amount used
+	st := tx.Rip7560TransactionData()
+	gasLimit := st.ValidationGas + st.PaymasterGas + params.Tx7560BaseGas
+	var (
+		lo uint64 // lowest-known gas limit where tx execution fails
+		hi uint64 // lowest-known gas limit where tx execution succeeds
+	)
+	// Determine the highest gas limit can be used during the estimation.
+	hi = opts.Header.GasLimit
+	if gasLimit >= params.TxGas {
+		hi = gasLimit
+	}
+	// Normalize the max fee per gas the call is willing to spend.
+	var feeCap *big.Int
+	if st.GasFeeCap != nil {
+		feeCap = st.GasFeeCap
+	} else {
+		feeCap = common.Big0
+	}
+	// Recap the highest gas limit with account's available balance.
+	if feeCap.BitLen() != 0 {
+		var payment common.Address
+		if len(st.PaymasterData) < 20 {
+			payment = *st.Sender
+		} else {
+			payment = common.BytesToAddress(st.PaymasterData[:20])
+		}
+		balance := opts.State.GetBalance(payment).ToBig()
+
+		allowance := new(big.Int).Div(balance, feeCap)
+
+		// If the allowance is larger than maximum uint64, skip checking
+		if allowance.IsUint64() && hi > allowance.Uint64() {
+			log.Debug("Gas estimation capped by limited funds", "original", hi, "balance", balance,
+				"maxFeePerGas", feeCap, "fundable", allowance)
+			hi = allowance.Uint64()
+		}
+	}
+	// Recap the highest gas allowance with specified gascap.
+	if gasCap != 0 && hi > gasCap {
+		log.Debug("Caller gas above allowance, capping", "requested", hi, "cap", gasCap)
+		hi = gasCap
+	}
+
+	// We first execute the transaction at the highest allowable gas limit, since if this fails we
+	// can return error immediately.
+	vpr, statedb, err := executeRip7560Validation(ctx, tx, opts, hi)
+	if err != nil {
+		if !errors.Is(err, vm.ErrOutOfGas) {
```

Comment:
Isn't the vm.ErrOutOfGas error case ruled out here due to the logic inside the executeRip7560Validation function?

Reply:
Oh yeah, you are right. I'll fix it, thanks.

Reply:
I fixed this to return gas error when vpr and err is both 0, since it's the first time trial with highest allowable gas limit.